src/rocksdb.nim

```diff
@@ -7,5 +7,224 @@
 #
 # at your option. This file may not be copied, modified, or distributed except according to those terms.
 
-import ./rocksdb_c
-export rocksdb_c
\ No newline at end of file
+import cpuinfo, options, ranges
+
+const useCApi = true
+
+when useCApi:
+  import rocksdb/librocksdb
+  export librocksdb
+
+  template managedResource(name) =
+    template freeResource(r: ptr `rocksdb name t`) =
+      `rocksdb name destroy`(r)
+
+  managedResource(WriteOptions)
+  managedResource(ReadOptions)
+
+  type
+    RocksPtr[T] = object
+      res: ptr T
+
+  import typetraits
+
+  when false:
+    # XXX: generic types cannot have destructors at the moment:
+    # https://github.com/nim-lang/Nim/issues/5366
+    proc `=destroy`*[T](rocksPtr: var RocksPtr[T]) =
+      freeResource rocksPtr.res
+
+  proc toRocksPtr[T](res: ptr T): RocksPtr[T] =
+    result.res = res
+
+  template initResource(resourceName): auto =
+    var p = toRocksPtr(`rocksdb resourceName create`())
+    # XXX: work-around the destructor issue above:
+    defer: freeResource p.res
+    p.res
+else:
+  {.error: "The C++ API of RocksDB is not supported yet".}
+
+  # The intention of this template is that it will hide the
+  # difference between the C and C++ APIs for objects such
+  # as Read/WriteOptions, which are allocated either on the
+  # stack or the heap.
+  template initResource(resourceName) =
+    var res = resourceName()
+    res
+
+type
+  # TODO: Replace this with a converter concept that will
+  # handle openarray[char] and openarray[byte] in the same way.
+  KeyValueType = openarray[byte]
+
+  RocksDBInstance* = object
+    db: ptr rocksdb_t
+    backupEngine: ptr rocksdb_backup_engine_t
+    options: ptr rocksdb_options_t
+
+  RocksDBResult*[T] = object
+    case ok*: bool
+    of true:
+      when T isnot void: value*: T
+    else:
+      error*: string
+
+template `$`*(s: RocksDBResult): string =
+  if s.ok:
+    $s.value
```

Comment:
if the s expression has side-effect, the side-effect will be called twice here;

Reply:
Hrm, this is a bit unfortunate. I wanted to avoid creating a string copy when the `RocksDBResult` is of type `string`, but most of the easy work-around will introduce similar copies. Changing to a proc for now (it's not a performance-critical API anyway).

Reply:
Ideally this should be catched by move semantics/sink in the new runtime.